runtime/src/chpl-topo.c

```diff
@@ -0,0 +1,284 @@
+  unsigned char* pCh;
+  size_t pgSize;
+  size_t pgMask;
+  unsigned char* pPgLo;
+  size_t nPages;
+  hwloc_obj_t numaObj;
+
+  _DBG_P("chpl_topo_setMemLocality(%p, %#zx, onlyIn=%s, doSub=%s, %d)\n",
+         p, size,
+         (onlyInside ? "T" : "F"), (doSubchunks ? "T" : "F"), (int) subloc);
+
+  if (!haveTopology) {
+    return;
+  }
+
+  pCh = (unsigned char*) p;
+  pgSize = chpl_getHeapPageSize();
+  pgMask = pgSize - 1;
+
+  if (onlyInside) {
+    pPgLo = round_up_to_mask_ptr(pCh, pgMask);
+    if (size < pPgLo - pCh)
+      nPages = 0;
+    else
+      nPages = round_down_to_mask(size - (pPgLo - pCh), pgMask) / pgSize;
+  } else {
+    pPgLo = round_down_to_mask_ptr(pCh, pgMask);
+    nPages = round_up_to_mask(size + (pCh - pPgLo), pgMask) / pgSize;
+  }
+
+  _DBG_P("    localize %p, %#zx bytes (%#zx pages)\n",
+         pPgLo, nPages * pgSize, nPages);
+
+  if (nPages == 0)
+    return;
+
+  if (doSubchunks) {
+    int i;
+    size_t pg;
+    size_t pgNext;
+
+    for (i = 0, pg = 0; i < numNumaDomains; i++, pg = pgNext) {
+      if (i == numNumaDomains - 1)
+        pgNext = nPages;
+      else
+        pgNext = 1 + (nPages * (i + 1) - 1) / numNumaDomains;
+      numaObj = hwloc_get_obj_by_depth(topology, numaLevel, i);
+      chpl_topo_setMemLocalityByPages(pPgLo + pg * pgSize,
+                                      (pgNext - pg) * pgSize,
+                                      numaObj->nodeset);
+    }
+  }
+  else if (subloc >= 0 && subloc < numNumaDomains) {
+    numaObj = hwloc_get_obj_by_depth(topology, numaLevel, subloc);
+    chpl_topo_setMemLocalityByPages(pPgLo, nPages * pgSize, numaObj->nodeset);
+  }
+}
+
+
+//
+// p must be page aligned and the page size must evenly divide size
+//
+static
+void chpl_topo_setMemLocalityByPages(unsigned char* p, size_t size,
+                                     hwloc_nodeset_t nodeset) {
+  const int flags = HWLOC_MEMBIND_MIGRATE | HWLOC_MEMBIND_STRICT;
+
+  if (!haveTopology) {
+    return;
+  }
+
+  if (!topoSupport->membind->set_area_membind)
+    return;
+
+  _DBG_P("hwloc_set_area_membind_nodeset(%p, %#zx, %d)\n", p, size,
+         (int) hwloc_bitmap_first(nodeset));
+  if (hwloc_set_area_membind_nodeset(topology, p, size,
+                                     nodeset, HWLOC_MEMBIND_BIND, flags)) {
+    report_error("hwloc_set_area_membind_nodeset()", errno);
+  }
+}
+
+
+c_sublocid_t chpl_topo_getMemLocality(void* p) {
+  const int flags = HWLOC_MEMBIND_STRICT;
+  hwloc_nodeset_t nodeset;
+  hwloc_membind_policy_t policy;
+  int node;
+
+  if (!haveTopology) {
+    return c_sublocid_any;
+  }
+
+  if (!topoSupport->membind->get_area_membind) {
+    return c_sublocid_any;
+  }
+
+  if ((nodeset = hwloc_bitmap_alloc()) == NULL) {
+    report_error("hwloc_bitmap_alloc()", errno);
+  }
+
+  if (hwloc_get_area_membind_nodeset(topology, p, 1,
+                                     nodeset, &policy, flags)) {
+    report_error("hwloc_get_area_membind_nodeset()", errno);
+  }
```

Comment:
An alternative would be to expand the error message for hwloc_get_area_membind_nodeset() failing, to print the value of the pointer passed to it.  Then it would at least be clear that the reason it failed was that it was passed a null pointer.